File: backend/src/controllers/backend/SessionController.ts
import { Response, Express } from "express";
import { Repository } from "typeorm";
import { Link } from "../../entity/Links";
import { getRepository, addMissingHTTP } from "../../utils";
import { isNullOrUndefined } from "util";
import { Session } from "../../entity/Session";

let linkRepo: Repository<Link>;
let sessionRepo: Repository<Session>;

const handleGetYear = async (req: any, res: Response) => {
    const re = await sessionRepo.query(
        `SELECT series AS month, COALESCE(clicks, 0) AS clicks
        FROM generate_series((date_trunc('month', current_date) - INTERVAL '12 months')::date, current_date, '1 month'::interval) AS series
        LEFT JOIN
        (SELECT COUNT(id) as clicks, date_trunc('month', timestamp) as month
        FROM session
        WHERE timestamp > (current_date - INTERVAL '12 months')
        GROUP BY date_trunc('month', timestamp)) AS stats
        ON stats.month = series
        ORDER BY month`
    );

    res.status(200).send(re);
};

const handleGetMonth = async (req: any, res: Response) => {
    const re = await sessionRepo.query(
        `SELECT series AS date, COALESCE(clicks, 0) AS clicks
        FROM generate_series((date_trunc('day', current_date) - INTERVAL '1 months')::date, current_date, '1 day'::interval) AS series
        LEFT JOIN
        (SELECT COUNT(id) as clicks, date_trunc('day', timestamp) as date
        FROM session
        WHERE date_trunc('month', timestamp) = date_trunc('month', current_date)
        GROUP BY date_trunc('day', timestamp)) AS stats
        ON stats.date = series
        ORDER BY date`
    );

    res.status(200).send(re);
};

const handleGetAverageHour = async (req: any, res: Response) => {
    const re = await sessionRepo.query(
        `SELECT series AS hour, COALESCE(clicks, 0) AS clicks
        FROM generate_series(1, 24, 1) AS series
        LEFT JOIN
        (SELECT COUNT(id) as clicks, date_part('hour', timestamp) as hour
        FROM session
        GROUP BY date_part('hour', timestamp)) AS stats
        ON stats.hour = series
        ORDER BY hour`
    );

    res.status(200).send(re);
};

const handleGetCountries = async (req: any, res: Response) => {
    const re = await sessionRepo.query(
        `SELECT COUNT(id) as clicks, country
        FROM session
        GROUP BY country`
    );

    res.status(200).send(re);
};

const sessionController = (app: Express) => {
    linkRepo = getRepository(Link);
    sessionRepo = getRepository(Session);
    app.get("/api/links/:id/sessions/year", handleGetYear);
    app.get("/api/links/:id/sessions/month", handleGetMonth);
    app.get("/api/links/:id/sessions/averageHour", handleGetAverageHour);
    app.get("/api/links/:id/sessions/countries", handleGetCountries);
};

export default sessionController;

File: go-backend/controllers/SessionController.go
package controllers

import (
	"fmt"
	"regexp"
	"time"

	"github.com/gin-contrib/sessions"
	"github.com/gin-gonic/gin"
)

var isAlphaNum = regexp.MustCompile(`^[a-zA-Z]+$`).MatchString

type year struct {
	Month  time.Time `json:"month"`
	Clicks int64     `json:"clicks"`
}

func handleGetYear(c *gin.Context) {
	session := sessions.Default(c)
	cid := session.Get("cid").(string)
	isAdmin := session.Get("isAdmin").(bool)
	id := c.Param("id")
	isOwner, link, err := isOwner(id, cid)
	if err != nil {
		c.AbortWithError(500, err)
		return
	}
	if !isOwner && !isAdmin {
		c.AbortWithStatus(403)
		return
	}
	var ret []year

	query := fmt.Sprintf(`SELECT series AS month, COALESCE(clicks, 0) AS clicks
	FROM generate_series((date_trunc('month', current_date) - INTERVAL '12 months')::date, current_date, '1 month'::interval) AS series
	LEFT JOIN
	(SELECT COUNT(id) as clicks, date_trunc('month', timestamp) as month
	FROM sessions
	WHERE timestamp > (current_date - INTERVAL '12 months') AND
	link_ref = %d
	GROUP BY date_trunc('month', timestamp)) AS stats
	ON stats.month = series
	ORDER BY month`, link.ID)

	res := db.Raw(query).Scan(&ret)

	if res.Error != nil {
		c.AbortWithError(500, res.Error)
	}

	c.JSON(200, ret)

}

type month struct {
	Date   time.Time `json:"date"`
	Clicks int64     `json:"clicks"`
}

func handleGetMonth(c *gin.Context) {
	session := sessions.Default(c)
	cid := session.Get("cid").(string)
	isAdmin := session.Get("isAdmin").(bool)
	id := c.Param("id")
	isOwner, link, err := isOwner(id, cid)
	if err != nil {
		c.AbortWithError(500, err)
		return
	}
	if !isOwner && !isAdmin {
		c.AbortWithStatus(403)
		return
	}
	var ret []month

	query := fmt.Sprintf(`SELECT series AS date, COALESCE(clicks, 0) AS clicks
	FROM generate_series((date_trunc('day', current_date) - INTERVAL '1 months')::date, current_date, '1 day'::interval) AS series
	LEFT JOIN
	(SELECT COUNT(id) as clicks, date_trunc('day', timestamp) as date
	FROM sessions
	WHERE date_trunc('month', timestamp) = date_trunc('month', current_date) AND
	link_ref = %d
	GROUP BY date_trunc('day', timestamp)) AS stats
	ON stats.date = series
	ORDER BY date`, link.ID)

	res := db.Raw(query).Scan(&ret)

	if res.Error != nil {
		c.AbortWithError(500, res.Error)
	}

	c.JSON(200, ret)

}

type day struct {
	Hour   int64 `json:"hour"`
	Clicks int64 `json:"clicks"`
}

func handleGetAverageHour(c *gin.Context) {
	session := sessions.Default(c)
	cid := session.Get("cid").(string)
	isAdmin := session.Get("isAdmin").(bool)
	id := c.Param("id")
	isOwner, link, err := isOwner(id, cid)
	if err != nil {
		c.AbortWithError(500, err)
		return
	}
	if !isOwner && !isAdmin {
		c.AbortWithStatus(403)
		return
	}
	var ret []day

	query := fmt.Sprintf(`SELECT series AS hour, COALESCE(clicks, 0) AS clicks
	FROM generate_series(1, 24, 1) AS series
	LEFT JOIN
	(SELECT COUNT(id) as clicks, date_part('hour', timestamp) as hour
	FROM sessions
	WHERE link_ref = %d
	GROUP BY date_part('hour', timestamp)) AS stats
	ON stats.hour = series
	ORDER BY hour`, link.ID)

	res := db.Raw(query).Scan(&ret)

	if res.Error != nil {
		c.AbortWithError(500, res.Error)
	}

	c.JSON(200, ret)

}

type countries struct {
	Clicks string    `json:"clicks"`
	Hour   time.Time `json:"country"`
}

func handleGetCountries(c *gin.Context) {
	session := sessions.Default(c)
	cid := session.Get("cid").(string)
	isAdmin := session.Get("isAdmin").(bool)
	id := c.Param("id")
	isOwner, link, err := isOwner(id, cid)
	if err != nil {
		c.AbortWithError(500, err)
		return
	}
	if !isOwner && !isAdmin {
		c.AbortWithStatus(403)
		return
	}
	var ret []countries

	query := fmt.Sprintf(`SELECT COUNT(id) as clicks, country
	FROM sessions
	WHERE link_ref = %d
	GROUP BY country`, link.ID)

	res := db.Raw(query).Scan(&ret)

	if res.Error != nil {
		c.AbortWithError(500, res.Error)
	}

	c.JSON(200, ret)

}

// RouteSessionController sets up routes for Session controller
func RouteSessionController(r *gin.RouterGroup) {
	r.GET("/:id/year", handleGetYear)
	r.GET("/:id/month", handleGetMonth)
	r.GET("/:id/averageHour", handleGetAverageHour)
	r.GET("/:id/countries", handleGetCountries)

}
